src/Orleans/Configuration/GlobalConfiguration.cs

```diff
@@ -217,21 +219,35 @@ public bool PrimaryNodeIsRequired
 
         #region MultiClusterNetwork
 
+        private string clusterId;
+
         /// <summary>
         /// Whether this cluster is configured to be part of a multicluster network
         /// </summary>
         public bool HasMultiClusterNetwork
         {
             get
             {
-                return !(string.IsNullOrEmpty(ClusterId));
+                return !(string.IsNullOrEmpty(this.clusterId));
             }
         }
 
         /// <summary>
         /// Cluster id (one per deployment, unique across all the deployments/clusters)
         /// </summary>
-        public string ClusterId { get; set; }
+        public string ClusterId
+        {
+            get
+            {
+                var configuredId = this.HasMultiClusterNetwork ? this.clusterId : this.DeploymentId;
+                return string.IsNullOrEmpty(configuredId) ? DefaultClusterId : configuredId;
+            }
+
+            set
+            {
+                this.clusterId = value;
+            }
+        }
```

Comment:
This logic needed to go somewhere. It previously lived in `Silo`, which is a bad place for it (i.e, consumers were accessing `Silo.CurrentSilo` to get to it).

I could put it somewhere else, like in an extension method or some new object